### PR TITLE
Fix crash when getting board.SPI outside the VM

### LINKED ISSUE #1753: hardfault on CPX build with displayio
getting hardfaults on boot and whenever i KeyboardInterrupt while trying to init a display on CPX built with displayio support

[files.zip](https://github.com/adafruit/circuitpython/files/3048941/files.zip)

### PR DESCRIPTION
**Pull request by tannewt**
If one of the default pins was already in use it would crash.

The internal API has been refined to allow us to get the value
without causing an init of the singleton.

Fixes #1753

### CODE REVIEW

**Comment by ladyada** [comment]
tested and it worked on my cpx!

File: supervisor/shared/board_busses.h [hunk 1]
```diff
@@ -29,7 +29,8 @@
 
 #include "py/obj.h"
 
-mp_obj_t board_i2c(void);
```
Comment (dhalbert) [inline supervisor/shared/board_busses.h]: The declaration of `board_i2c()` got deleted here, though `board_spi()` and `board_uart()` still exist. Was that deliberate?

File: shared-module/displayio/__init__.c [hunk 2]
```diff
@@ -190,7 +190,7 @@ void reset_displays(void) {
         if (((uint32_t) fourwire->bus) < ((uint32_t) &displays) ||
             ((uint32_t) fourwire->bus) > ((uint32_t) &displays + CIRCUITPY_DISPLAY_LIMIT)) {
             busio_spi_obj_t* original_spi = fourwire->bus;
-            if (original_spi == board_spi()) {
+            if (original_spi == common_hal_board_get_spi()) {
```
Comment (dhalbert) [inline shared-module/displayio/__init__.c]: Could you put in a comment here about the fact the `original_spi` or `common_hal_board_get_spi()` might be null, and if so why, and how that's handled? I'm having trouble figuring out the different possible cases.

File: supervisor/shared/board_busses.c [hunk 3]
```diff
@@ -71,23 +71,42 @@ MP_DEFINE_CONST_FUN_OBJ_0(board_i2c_obj, board_i2c);
...
+    return spi_singleton;
+}
+
+mp_obj_t common_hal_board_create_spi(void) {
+    if (spi_singleton != NULL) {
+        return spi_singleton;
+    }
+    busio_spi_obj_t *self = &spi_obj;
+    self->base.type = &busio_spi_type;
+    if (!common_hal_mcu_pin_is_free(DEFAULT_SPI_BUS_SCK) ||
```
Comment (dhalbert) [inline supervisor/shared/board_busses.c]: Could you add a comment? So, if the pins aren't free, we don't create, but just return null. Only `board_spi()` throws errors.

File: supervisor/shared/safe_mode.c [hunk 4]
```diff
@@ -77,7 +77,8 @@ safe_mode_t wait_for_safe_mode_reset(void) {
     return NO_SAFE_MODE;
 }
 
-void reset_into_safe_mode(safe_mode_t reason) {
+// Inline this so it's easy to break on it from GDB.
+void __attribute__((noinline,)) reset_into_safe_mode(safe_mode_t reason) {
```
Comment (dhalbert) [inline supervisor/shared/safe_mode.c]: "Don't inline this"?

File: supervisor/shared/board_busses.h (hunk 1 above)
Comment (tannewt) [reply supervisor/shared/board_busses.h]: Ya, I only switched one. Will push a larger refactor shortly.

**Comment by tannewt** [comment]
Ok, I've refactored everything so it touches many more files. Hopefully it all makes more sense. I did add a `gc_never_free` which is a catch all linked list from a root pointer for pointers that should live the whole rest of the heap. (Like SPI objects used by FourWire.)

**Comment by dhalbert** [comment]
> I did add a `gc_never_free` which is a catch all linked list from a root pointer for pointers that should live the whole rest of the heap. (Like SPI objects used by FourWire.)

That's a great idea!

File: shared-module/board/__init__.c [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+    MP_STATE_VM(shared_i2c_bus) = MP_OBJ_FROM_PTR(self);
+    return MP_STATE_VM(shared_i2c_bus);
+}
+
+// Statically allocate the SPI object so it can live past the end of the heap and into the next VM.
+// That way it can be used by built-in FourWire displays and be accessible through board.SPI().
+STATIC busio_spi_obj_t spi_obj;
+STATIC mp_obj_t spi_singleton = NULL;
+
+// TODO(tannewt): Move this to shared-bindings/board/__init__.c and corresponding shared-module.
```
Comment (dhalbert) [inline shared-module/board/__init__.c]: You can remove this TODO!

**Comment by dhalbert** [comment]
Bunch of builds failing :(

**Comment by tannewt** [comment]
Ok, I think this next run should pass. I fixed boards without shared busses, removed `audiobusio` from hallowing and removed `touchio` from pirkey to gain space back.

**Pull request merged**